spring_cloud/commons/client/discovery/composite.py

```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+# standard library
+from typing import List, Set
+
+# scip plugin
+from spring_cloud.commons.client import ServiceInstance
+
+__author__ = "Waterball (johnny850807@gmail.com)"
+__license__ = "Apache 2.0"
+
+# scip plugin
+from spring_cloud.commons.client.discovery.discovery_client import DiscoveryClient
+from spring_cloud.commons.utils.functional_operators import flat_map
+from spring_cloud.commons.utils.list_utils import not_none_nor_empty
+
+
+class CompositeDiscoveryClient(DiscoveryClient):
+    """
+    Composite pattern application:
+        aggregate the service sources from a list of discovery client
+    """
```

Comment:
What is the used of composite pattern in here ? composite pattern creates a structure of tree, but why do we need a tree here ?

Reply:
Since the source of the service instances will vary (statically from files, or from eureka registry, etc.)
> and yes it might be a **tree structure of sources**

This is the 'Chain Of Responsibility' (CoR) pattern actually (Though it's implemented in the Composite pattern way). CoR encapsulates that given a `serviceId`, which discovery client's will handle it.